File: libs/contracts/experiment_plan.py
"""
Pydantic schema for the FXLab Experiment Plan artifact (M2.C2).

Purpose:
    Parse and validate every ``experiment_plan.json`` artifact in the
    ``Strategy Repo/`` so that ``POST /runs/from-ir`` can accept a
    typed, immutable representation rather than raw dict access.
    This module is the schema-only layer — no business logic, no I/O,
    no run orchestration. The route layer (M2.C2 in
    ``services/api/routes/runs.py``) is responsible for resolving the
    referenced ``dataset_ref`` via :mod:`libs.strategy_ir.dataset_resolver`
    and forwarding the parsed plan to :class:`ResearchRunService`.

Responsibilities:
    - Define the root :class:`ExperimentPlan` model and every nested
      model that appears in any production experiment plan or in the
      canonical example (``User Spec/Algo Specfication and
      Examples/experiment_plan.example.json``).
    - Enforce the project-wide strict-frozen contract: every model has
      ``model_config = ConfigDict(extra='forbid', frozen=True)`` so any
      new field lands as a loud :class:`pydantic.ValidationError` rather
      than silently being dropped.
    - Pin ``schema_version`` to ``"0.1-inferred"`` and ``artifact_type``
      to ``"experiment_plan"`` (same strictness bar as
      :class:`libs.contracts.strategy_ir.StrategyIR` per workplan M1.A1).

Does NOT:
    - Read files from disk; the parser is the route handler's job.
    - Resolve ``dataset_ref`` against a dataset registry (Track E /
      M4.E3 will provide :class:`DatasetService`; until then the
      in-memory :class:`InMemoryDatasetResolver` is used by the route).
    - Build a :class:`ResearchRunConfig`; that mapping is performed by
      :class:`ResearchRunService.submit_from_ir`.

Dependencies:
    - Pydantic v2 only.

Schema reference:
    - ``User Spec/Algo Specfication and Examples/experiment_plan.example.json``
    - The five production experiment plans under ``Strategy Repo/``:
        * ``FX_TurnOfMonth_USDSeasonality_D1.experiment_plan.json``
        * ``FX_SingleAsset_MeanReversion_H1.experiment_plan.json``
        * ``FX_TimeSeriesMomentum_Breakout_D1.experiment_plan.json``
        * ``FX_DoubleBollinger_TrendZone.experiment_plan.json``
        * ``FX_MTF_DailyTrend_H1Pullback.experiment_plan.json``

Example::

    import json
    from libs.contracts.experiment_plan import ExperimentPlan

    with open("FX_DoubleBollinger_TrendZone.experiment_plan.json") as fh:
        plan = ExperimentPlan.model_validate(json.load(fh))

    assert plan.strategy_ref.strategy_name == "FX_DoubleBollinger_TrendZone"
    assert plan.data_selection.dataset_ref.startswith("fx-")
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# Reusable strict-frozen base
# ---------------------------------------------------------------------------


class _StrictFrozenModel(BaseModel):
    """
    Internal base providing the project-wide strict immutability config.

    Every experiment-plan model inherits from this so we never have to
    remember to repeat ``model_config = ConfigDict(extra='forbid',
    frozen=True)`` -- forgetting it on even one nested model would
    silently allow extra fields and mutation, defeating the schema's
    whole job. Mirrors the pattern in
    :mod:`libs.contracts.strategy_ir`.
    """

    model_config = ConfigDict(extra="forbid", frozen=True)


# ---------------------------------------------------------------------------
# strategy_ref
# ---------------------------------------------------------------------------


class StrategyRef(_StrictFrozenModel):
    """
    Reference to the strategy this experiment plan exercises.

    The ``strategy_name`` matches ``StrategyIR.metadata.strategy_name``
    in the matching ``strategy_ir.json`` artifact, and the version
    pin is independent of the strategy IR version because experiment
    plans evolve faster than the strategies they target.
    """

    strategy_name: str = Field(..., min_length=1, max_length=256)
    strategy_version: str = Field(..., min_length=1, max_length=64)


# ---------------------------------------------------------------------------
# run_metadata
# ---------------------------------------------------------------------------


class RunMetadata(_StrictFrozenModel):
    """
    Bookkeeping fields for a research run instance.

    ``random_seed`` is required because every research run in FXLab
    must be byte-deterministic. ``run_purpose`` is free text: the
    canonical example uses ``"baseline_research_and_parameter_search"``
    and downstream stratification labels treat it as opaque.
    """

    run_purpose: str = Field(..., min_length=1, max_length=256)
    owner: str = Field(..., min_length=1, max_length=256)
    random_seed: int


# ---------------------------------------------------------------------------
# data_selection
# ---------------------------------------------------------------------------


class DataSelection(_StrictFrozenModel):
    """
    Dataset, spread-dataset, and calendar references for the run.

    ``dataset_ref`` is the load-bearing field: the route handler
    resolves it via :class:`DatasetResolverInterface` (M4.E3 will swap
    in the real :class:`DatasetService` adapter; until then the
    in-memory map is used). ``spread_dataset_ref`` and ``calendar_ref``
    are passed straight through to the engine config without
    resolution at this tranche.
    """

    dataset_ref: str = Field(..., min_length=1, max_length=256)
    dataset_version: str = Field(..., min_length=1, max_length=64)
    spread_dataset_ref: str = Field(..., min_length=1, max_length=256)
    calendar_ref: str = Field(..., min_length=1, max_length=256)


# ---------------------------------------------------------------------------
# cost_models
# ---------------------------------------------------------------------------


class CostModels(_StrictFrozenModel):
    """
    Pinned references for commission, slippage, and swap models.

    All three refs are opaque strings at the schema layer; the
    backtest engine resolves them against its own registries.
    """

    commission_model_ref: str = Field(..., min_length=1, max_length=256)
    slippage_model_ref: str = Field(..., min_length=1, max_length=256)
    swap_model_ref: str = Field(..., min_length=1, max_length=256)


# ---------------------------------------------------------------------------
# splits
# ---------------------------------------------------------------------------


class DateRange(_StrictFrozenModel):
    """
    A ``[start, end]`` ISO-8601 date range used by every split.

    Stored as plain strings rather than :class:`datetime.date` because
    the IR is consumed by both Python and (eventually) JS clients;
    keeping the wire shape as strings avoids timezone ambiguity. The
    engine layer is responsible for parsing.
    """

    start: str = Field(..., min_length=1, max_length=32)
    end: str = Field(..., min_length=1, max_length=32)


class Splits(_StrictFrozenModel):
    """
    Three-way temporal split: in-sample / out-of-sample / holdout.

    All three splits are required across every production plan we
    have surveyed. The engine enforces non-overlap; the schema does
    not, because compliance reporting cares about the originally
    declared boundaries even if they overlap.
    """

    in_sample: DateRange
    out_of_sample: DateRange
    holdout: DateRange


# ---------------------------------------------------------------------------
# validation
# ---------------------------------------------------------------------------


class WalkForwardSpec(_StrictFrozenModel):
    """
    Walk-forward validation parameters.

    Disabled when ``enabled=False``; remaining fields are still
    validated for shape so the engine config layer never has to
    check for ``None``.
    """

    enabled: bool
    train_window_months: int = Field(..., gt=0)
    test_window_months: int = Field(..., gt=0)
    step_months: int = Field(..., gt=0)


class MonteCarloSpec(_StrictFrozenModel):
    """
    Monte-Carlo validation parameters.

    ``method`` is free-text at this tranche; the engine layer maps
    known methods (e.g. ``"trade_sequence_resampling"``) to concrete
    resamplers and rejects unknown values.
    """

    enabled: bool
    iterations: int = Field(..., gt=0)
    method: str = Field(..., min_length=1, max_length=128)


class RegimeSegmentationSpec(_StrictFrozenModel):
    """
    Regime-segmentation validation parameters.

    ``dimensions`` is a free-form list of regime tags; the engine
    layer matches them against its registered regime classifiers.
    """

    enabled: bool
    dimensions: list[str] = Field(..., min_length=1)


class Validation(_StrictFrozenModel):
    """
    Aggregated validation spec covering all three sub-frameworks.

    Each sub-spec is required because every production plan provides
    all three; the engine layer honours each section's ``enabled``
    flag independently.
    """

    walk_forward: WalkForwardSpec
    monte_carlo: MonteCarloSpec
    regime_segmentation: RegimeSegmentationSpec


# ---------------------------------------------------------------------------
# ranking
# ---------------------------------------------------------------------------


class Ranking(_StrictFrozenModel):
    """
    Primary + secondary ranking metrics for the experiment.

    ``primary_metric`` is the single metric used for parameter-search
    selection; ``secondary_metrics`` are reported alongside but do
    not break ties.
    """

    primary_metric: str = Field(..., min_length=1, max_length=128)
    secondary_metrics: list[str] = Field(..., min_length=1)


# ---------------------------------------------------------------------------
# acceptance_thresholds
# ---------------------------------------------------------------------------


class AcceptanceThresholds(_StrictFrozenModel):
    """
    Hard acceptance thresholds used by the readiness report.

    A run that fails to clear any threshold is flagged as
    NON-PROMOTABLE. The schema enforces sensible numeric guards
    (positive trade count, non-negative drawdown) but leaves
    semantic comparison to the engine.
    """

    min_trade_count: int = Field(..., ge=0)
    min_profit_factor: float = Field(..., ge=0.0)
    max_drawdown_pct: float = Field(..., ge=0.0)
    min_out_of_sample_sharpe: float
    min_holdout_profit_factor: float = Field(..., ge=0.0)


# ---------------------------------------------------------------------------
# outputs
# ---------------------------------------------------------------------------


class Outputs(_StrictFrozenModel):
    """
    Required artifact list and persistence flag.

    ``required`` lists the artifacts the engine must emit (e.g.
    ``"trade_blotter"``, ``"equity_curve"``); ``persist_artifacts``
    controls whether they are written to durable storage.
    """

    required: list[str] = Field(..., min_length=1)
    persist_artifacts: bool


# ---------------------------------------------------------------------------
# Root model
# ---------------------------------------------------------------------------


class ExperimentPlan(_StrictFrozenModel):
    """
    The full experiment-plan artifact.

    Mirrors the top-level shape of every ``*.experiment_plan.json``
    file under ``Strategy Repo/``. Every field is either required or
    populates with a stable default; ``extra='forbid'`` ensures any
    drift surfaces as a :class:`pydantic.ValidationError` at parse
    time rather than silently corrupting downstream behaviour.

    Attributes:
        schema_version: Always ``"0.1-inferred"`` until a versioned
            schema replaces it.
        artifact_type: Always ``"experiment_plan"``.
        strategy_ref: Pointer to the target strategy IR.
        run_metadata: Owner, purpose, RNG seed.
        data_selection: Dataset, spread, calendar refs.
        cost_models: Commission/slippage/swap model refs.
        splits: In-sample / out-of-sample / holdout date ranges.
        validation: Walk-forward / Monte-Carlo / regime configs.
        ranking: Primary + secondary metric definitions.
        acceptance_thresholds: Hard pass/fail thresholds.
        outputs: Required artifact list + persistence flag.
        notes: Free-text annotations (always a list of strings in
            production plans; the canonical example matches).
    """

    schema_version: Literal["0.1-inferred"]
    artifact_type: Literal["experiment_plan"]
    strategy_ref: StrategyRef
    run_metadata: RunMetadata
    data_selection: DataSelection
    cost_models: CostModels
    splits: Splits
    validation: Validation
    ranking: Ranking
    acceptance_thresholds: AcceptanceThresholds
    outputs: Outputs
    notes: list[str] = Field(default_factory=list)


__all__ = [
    "AcceptanceThresholds",
    "CostModels",
    "DataSelection",
    "DateRange",
    "ExperimentPlan",
    "MonteCarloSpec",
    "Outputs",
    "Ranking",
    "RegimeSegmentationSpec",
    "RunMetadata",
    "Splits",
    "StrategyRef",
    "Validation",
    "WalkForwardSpec",
]

File: frontend/src/types/experiment_plan.ts
/**
 * ExperimentPlan TypeScript types.
 *
 * Purpose:
 *   Mirror of ``libs/contracts/experiment_plan.py`` (M2.C2). Every nested
 *   model in the Pydantic schema has a matching ``interface`` here so the
 *   ``RunBacktestModal`` and the ``api/runs.ts`` client can build a
 *   wire-compatible payload for ``POST /runs/from-ir`` without round-tripping
 *   through ``unknown``.
 *
 * Responsibilities:
 *   - Export structural types matching the Pydantic ExperimentPlan exactly.
 *   - Provide a lightweight runtime validator (``validateExperimentPlan``)
 *     that mirrors the Pydantic invariants the modal can surface inline
 *     (required strings non-empty, positive integers, etc.). The backend is
 *     still the authoritative validator (it owns Pydantic) — this is a
 *     fast-feedback gate so users do not have to wait for a 422 round trip
 *     to learn that ``train_window_months`` must be > 0.
 *   - Provide a ``buildEmptyExperimentPlan()`` factory that returns a
 *     plan skeleton with all required fields set to safe placeholder
 *     values so the form can bind ``value=`` inputs without juggling
 *     ``undefined``.
 *
 * Does NOT:
 *   - Make API calls (see ``api/runs.ts``).
 *   - Resolve dataset references (the backend route does that).
 *   - Implement the full Pydantic strict-frozen guarantee — TypeScript
 *     interfaces cannot reject extra fields at runtime; the backend will
 *     return 422 for unknown fields.
 *
 * Schema reference:
 *   - libs/contracts/experiment_plan.py (M2.C2)
 *   - Strategy Repo/**\/*.experiment_plan.json
 */

// ---------------------------------------------------------------------------
// Nested types (one per Pydantic model in libs/contracts/experiment_plan.py)
// ---------------------------------------------------------------------------

export interface StrategyRef {
  strategy_name: string;
  strategy_version: string;
}

export interface RunMetadata {
  run_purpose: string;
  owner: string;
  random_seed: number;
}

export interface DataSelection {
  dataset_ref: string;
  dataset_version: string;
  spread_dataset_ref: string;
  calendar_ref: string;
}

export interface CostModels {
  commission_model_ref: string;
  slippage_model_ref: string;
  swap_model_ref: string;
}

export interface DateRange {
  start: string;
  end: string;
}

export interface Splits {
  in_sample: DateRange;
  out_of_sample: DateRange;
  holdout: DateRange;
}

export interface WalkForwardSpec {
  enabled: boolean;
  train_window_months: number;
  test_window_months: number;
  step_months: number;
}

export interface MonteCarloSpec {
  enabled: boolean;
  iterations: number;
  method: string;
}

export interface RegimeSegmentationSpec {
  enabled: boolean;
  dimensions: string[];
}

export interface Validation {
  walk_forward: WalkForwardSpec;
  monte_carlo: MonteCarloSpec;
  regime_segmentation: RegimeSegmentationSpec;
}

export interface Ranking {
  primary_metric: string;
  secondary_metrics: string[];
}

export interface AcceptanceThresholds {
  min_trade_count: number;
  min_profit_factor: number;
  max_drawdown_pct: number;
  min_out_of_sample_sharpe: number;
  min_holdout_profit_factor: number;
}

export interface Outputs {
  required: string[];
  persist_artifacts: boolean;
}

export interface ExperimentPlan {
  schema_version: "0.1-inferred";
  artifact_type: "experiment_plan";
  strategy_ref: StrategyRef;
  run_metadata: RunMetadata;
  data_selection: DataSelection;
  cost_models: CostModels;
  splits: Splits;
  validation: Validation;
  ranking: Ranking;
  acceptance_thresholds: AcceptanceThresholds;
  outputs: Outputs;
  notes: string[];
}

// ---------------------------------------------------------------------------
// Validation
// ---------------------------------------------------------------------------

/**
 * Field-level validation errors, keyed by dotted path matching the
 * Pydantic field path. The modal renders the message inline next to the
 * matching input so the user can correct one field at a time.
 *
 * Example keys:
 *   - "strategy_ref.strategy_name"
 *   - "splits.in_sample.start"
 *   - "validation.walk_forward.train_window_months"
 */
export type ExperimentPlanErrors = Record<string, string>;

/** ISO-8601 date check: ``YYYY-MM-DD`` only — calendar dates, no time. */
const ISO_DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

/**
 * Validate a hand-edited :class:`ExperimentPlan` against the same shape
 * the backend Pydantic validator enforces.
 *
 * Mirrors the most important invariants from
 * ``libs/contracts/experiment_plan.py``:
 *   - Required strings are non-empty.
 *   - All ``*_window_months`` / ``step_months`` / ``iterations`` > 0.
 *   - ``acceptance_thresholds`` numerics that have ``ge=0`` constraints.
 *   - ``splits`` dates parse as ISO-8601 ``YYYY-MM-DD``.
 *   - At least one item in lists declared with ``min_length=1``.
 *
 * Args:
 *   plan: The plan candidate from the modal form state.
 *
 * Returns:
 *   A flat error map. Empty object means the plan passes the
 *   client-side gate (the backend will still validate).
 */
export function validateExperimentPlan(plan: ExperimentPlan): ExperimentPlanErrors {
  const errors: ExperimentPlanErrors = {};

  const requireString = (path: string, value: string): void => {
    if (!value || value.trim() === "") {
      errors[path] = "Required";
    }
  };

  const requirePositiveInt = (path: string, value: number): void => {
    if (!Number.isInteger(value) || value <= 0) {
      errors[path] = "Must be a positive integer";
    }
  };

  const requireNonNegative = (path: string, value: number): void => {
    if (!Number.isFinite(value) || value < 0) {
      errors[path] = "Must be zero or greater";
    }
  };

  const requireDate = (path: string, value: string): void => {
    if (!ISO_DATE_RE.test(value)) {
      errors[path] = "Must be YYYY-MM-DD";
    }
  };

  // strategy_ref
  requireString("strategy_ref.strategy_name", plan.strategy_ref.strategy_name);
  requireString("strategy_ref.strategy_version", plan.strategy_ref.strategy_version);

  // run_metadata
  requireString("run_metadata.run_purpose", plan.run_metadata.run_purpose);
  requireString("run_metadata.owner", plan.run_metadata.owner);
  if (!Number.isInteger(plan.run_metadata.random_seed)) {
    errors["run_metadata.random_seed"] = "Must be an integer";
  }

  // data_selection
  requireString("data_selection.dataset_ref", plan.data_selection.dataset_ref);
  requireString("data_selection.dataset_version", plan.data_selection.dataset_version);
  requireString("data_selection.spread_dataset_ref", plan.data_selection.spread_dataset_ref);
  requireString("data_selection.calendar_ref", plan.data_selection.calendar_ref);

  // cost_models
  requireString("cost_models.commission_model_ref", plan.cost_models.commission_model_ref);
  requireString("cost_models.slippage_model_ref", plan.cost_models.slippage_model_ref);
  requireString("cost_models.swap_model_ref", plan.cost_models.swap_model_ref);

  // splits
  requireDate("splits.in_sample.start", plan.splits.in_sample.start);
  requireDate("splits.in_sample.end", plan.splits.in_sample.end);
  requireDate("splits.out_of_sample.start", plan.splits.out_of_sample.start);
  requireDate("splits.out_of_sample.end", plan.splits.out_of_sample.end);
  requireDate("splits.holdout.start", plan.splits.holdout.start);
  requireDate("splits.holdout.end", plan.splits.holdout.end);

  // validation.walk_forward — when enabled, all windows must be > 0.
  // The backend uses ``gt=0`` regardless of the enabled flag, so we
  // mirror that strictly to avoid a 422 surprise.
  requirePositiveInt(
    "validation.walk_forward.train_window_months",
    plan.validation.walk_forward.train_window_months,
  );
  requirePositiveInt(
    "validation.walk_forward.test_window_months",
    plan.validation.walk_forward.test_window_months,
  );
  requirePositiveInt(
    "validation.walk_forward.step_months",
    plan.validation.walk_forward.step_months,
  );

  // validation.monte_carlo
  requirePositiveInt("validation.monte_carlo.iterations", plan.validation.monte_carlo.iterations);
  requireString("validation.monte_carlo.method", plan.validation.monte_carlo.method);

  // validation.regime_segmentation — at least one dimension required.
  if (plan.validation.regime_segmentation.dimensions.length === 0) {
    errors["validation.regime_segmentation.dimensions"] = "At least one dimension required";
  }

  // ranking
  requireString("ranking.primary_metric", plan.ranking.primary_metric);
  if (plan.ranking.secondary_metrics.length === 0) {
    errors["ranking.secondary_metrics"] = "At least one secondary metric required";
  }

  // acceptance_thresholds — ge=0 on most, free sign on min_out_of_sample_sharpe.
  requireNonNegative(
    "acceptance_thresholds.min_trade_count",
    plan.acceptance_thresholds.min_trade_count,
  );
  if (
    !Number.isInteger(plan.acceptance_thresholds.min_trade_count) ||
    plan.acceptance_thresholds.min_trade_count < 0
  ) {
    errors["acceptance_thresholds.min_trade_count"] = "Must be a non-negative integer";
  }
  requireNonNegative(
    "acceptance_thresholds.min_profit_factor",
    plan.acceptance_thresholds.min_profit_factor,
  );
  requireNonNegative(
    "acceptance_thresholds.max_drawdown_pct",
    plan.acceptance_thresholds.max_drawdown_pct,
  );
  if (!Number.isFinite(plan.acceptance_thresholds.min_out_of_sample_sharpe)) {
    errors["acceptance_thresholds.min_out_of_sample_sharpe"] = "Must be a number";
  }
  requireNonNegative(
    "acceptance_thresholds.min_holdout_profit_factor",
    plan.acceptance_thresholds.min_holdout_profit_factor,
  );

  // outputs — at least one required artifact.
  if (plan.outputs.required.length === 0) {
    errors["outputs.required"] = "At least one required artifact";
  }

  return errors;
}

// ---------------------------------------------------------------------------
// Empty / preset factory
// ---------------------------------------------------------------------------

/**
 * Build an empty :class:`ExperimentPlan` with shape-correct defaults.
 *
 * The form binds ``value=`` directly against this object; using empty
 * strings (rather than ``undefined``) keeps the inputs controlled and
 * lets the modal mount without flicker. The validator will reject the
 * empty plan until the user fills it in.
 */
export function buildEmptyExperimentPlan(): ExperimentPlan {
  return {
    schema_version: "0.1-inferred",
    artifact_type: "experiment_plan",
    strategy_ref: { strategy_name: "", strategy_version: "" },
    run_metadata: { run_purpose: "", owner: "", random_seed: 0 },
    data_selection: {
      dataset_ref: "",
      dataset_version: "",
      spread_dataset_ref: "",
      calendar_ref: "",
    },
    cost_models: {
      commission_model_ref: "",
      slippage_model_ref: "",
      swap_model_ref: "",
    },
    splits: {
      in_sample: { start: "", end: "" },
      out_of_sample: { start: "", end: "" },
      holdout: { start: "", end: "" },
    },
    validation: {
      walk_forward: {
        enabled: false,
        train_window_months: 12,
        test_window_months: 3,
        step_months: 1,
      },
      monte_carlo: { enabled: false, iterations: 100, method: "trade_sequence_resampling" },
      regime_segmentation: { enabled: false, dimensions: [] },
    },
    ranking: { primary_metric: "", secondary_metrics: [] },
    acceptance_thresholds: {
      min_trade_count: 0,
      min_profit_factor: 0,
      max_drawdown_pct: 0,
      min_out_of_sample_sharpe: 0,
      min_holdout_profit_factor: 0,
    },
    outputs: { required: [], persist_artifacts: true },
    notes: [],
  };
}
